### DISCUSSION TITLE
How to make LocalStorage reactive?

### DISCUSSION THREAD

**Discussion by Serhioromano** [quasarframework/quasar#14280] [General - Components / Directives / etc] [answered]
My app is like a browser. Each page is opened in a new tab. I wanted to use Local Storage to remember those tabs. Here is my helper.

```js
function useTabs() {
    interface Tab {
        title: string;
        route: string;
        icon: string;
    }
    const deleteTab = (id: string) => {
        const tabs = <Tab[]>LocalStorage.getItem("tabs");
        tabs.forEach((val, index) => {
            if (val.route === id) {
                tabs.splice(index, 1);
            }
        });
        LocalStorage.set("tabs", tabs);
    };
    const addTab = (Tab: Obj) => {
        let tabs = <Tab[]>LocalStorage.getItem("tabs");
        if (tabs === null) {
            tabs = [];
        }
        let toAdd = true;
        tabs.forEach(val => {
            if (val.route === Tab.route) {
                toAdd = false;
            }
        });
        if (toAdd) {
            tabs.push({
                route: Tab.route,
                title: Tab.title,
                icon: Tab.icon,
            });
            LocalStorage.set("tabs", tabs);
        }
    };
    return {
        deleteTab,
        addTab,
    };
}
```

Then in main layout

```html
<q-tabs v-model="tabbar" v-if="tabsList.length > 0" inline-label align="left">
    <q-route-tab
        v-for="tab in tabsList"
        no-caps
        :key="tab.route" :to="tab.route" :name="tab.title" :icon="tab.icon" :label="tab.title"
    />
</q-tabs>
```

and script

```js
setup() {
    const $q = useQuasar();
    const tabsList = ref($q.localStorage.getItem("tabs"));
    const {deleteTab} = useTabs();

    return {
        tabsList,
        deleteTab,
    };
},
```

If I look in browser inspector, I can see tab is added when I open document, but I must refresh browser to see those tabs. How do I make that when it is added it immediately show tabs?

I created [sandbox example](https://codesandbox.io/s/modest-violet-wquwpf?file=/src/pages/Gost.vue).

**Comment by yusufkandemir** [comment]
`getItem` returns a static value, wrapping it with `ref` makes it the initial value of that ref. So, it's basically like `ref(0)`, which will not get updated unless you update it. So, what you are experiencing is expected behavior. To make it reactive, you would have to use it as the initial value, but keep manipulating the variable as if there is no local storage, but watch that variable and keep local storage in sync so when you refresh the page the initial value is what you expect.

Combined with some refactoring advice, here is the code I would recommend:

```ts
interface Tab {
  title: string;
  route: string;
  icon: string;
}

function useTabs() {
  const tabs = ref(LocalStorage.getItem<Tab[]>("tabs") || []);
  watch(tabs, (value) => {
    LocalStorage.set("tabs", value);
  });
  
  function deleteTab (id: string) {
    const index = tabs.value.findIndex(({ route }) => route === id);
    if (index === -1) {
      throw new Error("Tab not found!");
    }
    tabs.value.splice(index, 1);
  }

  function addTab (tab: Tab) {
    const exists = tabs.value.some(({ route }) => route === tab.route);
    if (exists) {
      return;
    }

    tabs.value.push(tab);
  }

  return {
    tabs,
    // ...
  }
}

// In setup:
const { tabs, deleteTab } = useTabs();
```

**Reply by Serhioromano** [reply]
I see. But question about `watch()`. It is not a mixin or component. It is a simple class so to say, how watch is going to work? I [used this](https://www.syncfusion.com/blogs/post/can-the-composition-api-replace-vue-mixins.aspx#:~:text=Role%20of%20Composition%20API%20in,plug%2Din%20in%20Vue%202.) article to create component. As I understand there is not `setup()` or `watch()` interface there.

**Reply by Serhioromano** [reply]
I have added sandbox example [link](https://codesandbox.io/s/modest-violet-wquwpf?file=/src/pages/Gost.vue) to original post.

**Reply by yusufkandemir** [reply]
First of all, if you don't know already, you are using Composition API. The "unit" you are using here (`useXYZ`) is often called a composable. Composables are just functions, they are not mixins or components, although they can be used to replace them fully or partially. Mixins are meant to be used with Options API, so they are not relevant in your case. Composables are generally meant to be called inside `setup()`, which you are doing already, but can actually be used anywhere. So, it's normal stuff, there is nothing to worry about. Vue's reactivity functions can be used anywhere, you are not limited to setup, composables, etc. I would suggest reading some docs for more information.

Don't forget to choose Composition API to get the relevant content.
![image](https://user-images.githubusercontent.com/6266078/187721680-639b9073-6a88-42c4-bb3b-945c5dc7a919.png)

https://vuejs.org/guide/essentials/reactivity-fundamentals.html
https://vuejs.org/guide/extras/reactivity-in-depth.html

**Comment by metalsadman** [comment]
There are vuex/pinia plugins out there.

**Reply by TobyMosque** [reply]
https://prazdevs.github.io/pinia-plugin-persistedstate/frameworks/quasar.html#usage-with-quasar-2

**Comment by dongwa** [comment, marked as answer]
Meby you can see [vueuse for Reactive LocalStorage](https://vueuse.org/core/useLocalStorage/)

**Reply by Serhioromano** [reply]
Thank you! I did not know about this library. It addresses so many things I need!
